### ISSUE TITLE
Réunion - 03/08/2015

### ISSUE THREAD

**Issue by pierre-H** [rcharb1/optimiseur-coupe#2]
Salut @rcharb1,
J'ai mis dans le dossier Doc une proposition du protocole à adopter.
De plus, j'ai ajouté un dossier client qui contiendrait le client. Cela ne sert à rien de le compiler car je n'ai que : 
- ajouté le support du protocole ODC (Optimiseur De Coupe) décrit dans Doc/Communication.md (dossier Réseau)
- Supprimé le moteur

Je n'ai pas ajouté la véritable partie réseau en TCP et la connexion entre le protocole ODC et le client.

Dis-moi quand tu veux que l'on commence la discussion (voir un horaire). J'ai un accès Internet mais pas très puisant, donc ne t'étonne pas si j'ai du mal à répondre sur le champ.

Si tu peux lire avant le document Doc/Communication.md et me dire ce que tu en penses...

**Comment by pierre-H** [comment]
Plusieurs questions me viennent à l'esprit :
- Est-ce que l'on choisit un port par défaut non modifiable ou on laisse les utilisateurs choisir lors du lancement du client et du serveur ?
- pour le client, est-ce que avant l'ouverture de la fenêtre principale on affiche une petite fenêtre demandant l'adresse IP (et le port) ou on ajoute des champs adresse IP (et port) dans le formulaire de saisie ?

**Comment by rcharb1** [comment]
Désolée de faire faux bond... Est-ce que ce serait dérangeant de repousser à demain soir ? J'ai été extrêmement occupée ces derniers jours et je n'ai pas eu beaucoup de temps pour travailler dessus...
Si ça pose le moindre problème, je peux encore travailler dans les deux heures qui viennent et qu'on se donne RDV pour 9h...
Dis moi ce que tu préfères... Encore désolée...
Je regarde maintenant les docs mis en ligne.

**Comment by pierre-H** [comment]
On peut faire ça demain oui. Quelle heure ?
Je viens de modifier Communication.md . Je vais continuer à travailler sur le client...

**Comment by rcharb1** [comment]
8h ? Je viens de regarder communication.md. C'est vraiment très bien! Et moi, je fais le serveur d'ici demain si D' veut... Encore désolée...

**Comment by pierre-H** [comment]
Très bien 8h (heure française n'est-ce pas ? )
Merci beaucoup. Si tu remarques une erreur ou un manque, n'hésite pas à le modifier.
Pas de problème. Moi c'était le mois dernier où j'étais trop occupé.
Il y a de toutes les façons beaucoup de boulots pour le client (notamment la partie résultats texte et graphique qui utilisait tes structures de contrôles et qui maintenant va utiliser des structures de Qt ou des classes personnalisées simples).
Si tu peux aussi réfléchir ton avis sur les 2 questions posées là haut, ce serait top !

**Comment by rcharb1** [comment]
Pour les deux questions : 
1. Je pense qu'un port par défaut est une bonne idée, maisil faut que le programme ait la souplesse de le changer si ce port est utilisé dans la machine en question.
2. Je pense que tout dans le formulaire de saisie fait un peu plus propre...

**Comment by pierre-H** [comment]
Je viens de mettre une nouvelle version du programme qui demande l'adresse IP et le port lors du lancement du programme. Je viens seulement de voir tes réponses. Quel port par défaut du coup ?
J'ai aussi modifié le protocole.

**Comment by pierre-H** [comment]
Je viens de voir ton mail : c'est une bonne nouvelle en soi. Fais comme tu peux. De mon côté, le gros travail sur le client a été fait. Si tu ne penses pas pouvoir faire le serveur, je le ferai. Pas de souci.

**Comment by rcharb1** [comment]
Oui, mais ce n'est pas juste, le projet doit être fait à 2...

**Comment by rcharb1** [comment]
(désolée pour le retard... j'ai l'impression d'un tremblement de terre qui m'est tombé dessus)

**Comment by pierre-H** [comment]
Je ne pense pas : le projet doit normalement être fait seul, c'est JM qui a donné son autorisation pour qu'on le fasse à deux.

Pas de souci pour le retard. Au stage, c'était 35h par semaine avec 2h40 de trajets : j'étais noyé !
Je comprend tout à fait cela. Fais ce que tu peux.

**Comment by rcharb1** [comment]
En fait, ça va plutôt être 50 heures par semaine avec 2 h de trajet...

**Comment by rcharb1** [comment]
Donc qu'on fasse chacun de son côté ?

**Comment by rcharb1** [comment]
Peut-être que je peux y arriver parce que j'ai le vendredi de libre...

**Comment by pierre-H** [comment]
C'est beaucoup en effet.
Je pense qu'il faut faire le maximum pour qu'on fasse ensemble le projet.
Personnellement, j'ai mis une journée pour le client. Je pense que ce sera la même chose pour le client.

**Comment by pierre-H** [comment]
Pour le serveur, pardon...

**Comment by rcharb1** [comment]
Oui, je pense qu'en regardant ce que tu as déjà fait, ça peut vraiment me faciliter la vie pour le serveur et je peux y arriver...

**Comment by rcharb1** [comment]
Maintenant, faisons comme si j'avais fait ce que je devais faire et voyons de quoi il faut qu'on discute...

**Comment by pierre-H** [comment]
Oui. Il y a [ce cours](https://openclassrooms.com/courses/programmez-avec-le-langage-c/communiquer-en-reseau-avec-son-programme) qui m'a beaucoup aidé aussi.
Oui.

**Comment by rcharb1** [comment]
ça, j'ai lu, il faut que je l'étudie plus en détails pour bien l'intégrer, mais il semble effectivement contenir le code de base.
Finalement, on devrait peut-être faire comme lui de mettre une fenêtre pour le serveur...

**Comment by rcharb1** [comment]
Qu'en est-il de la version plus compliquée du programme dont on a dit qu'il fallait parler ?

**Comment by pierre-H** [comment]
Concernant l'entrée de l'adresse IP et du port, actuellement le programme demande l'IP puis le port. Si la connexion est réussi, on affiche la fenêtre principale. Sinon, on propose soit la fermeture du programme soit de réessayer. Les erreurs sont affichés, si possible en français.
J'ai fait ainsi car ce n'est pas logique je trouve d'afficher le formulaire si on n'a pas de connexion avec un serveur-moteur. Le premier est conditionné par le 2e.
Mais on peut tout avoir dans la même fenêtre.
Il faut que l'on se mette d'accord sur le port par défaut.

**Comment by pierre-H** [comment]
Pour la fenêtre pour le serveur, on peut, mais il faut donc donner la possibilité de choisir le port contrairement à ce qu'il a fait.

**Comment by rcharb1** [comment]
Je viens de voir, c'est très clean. Bravo !

**Comment by pierre-H** [comment]
Merci :)

**Comment by pierre-H** [comment]
Depuis l'optimiseur de coupe original, j'essaye de coder toujours avec des commentaires et de garder les même règles de syntaxe comme indiqué par Gilles Bernard.

**Comment by rcharb1** [comment]
je propose qu'on choisisse un numéro légèrement plus grand que 1024 et qu'on définisse un random (1025, 5000) pour le cas où il est pris

**Comment by rcharb1** [comment]
Je regarderai donc ta manière de coder pour m'y conformer

**Comment by pierre-H** [comment]
Pourquoi ne pas donner le choix à l'utilisateur du port ?

**Comment by rcharb1** [comment]
Parce que mamie va chercher le mot port dans un dictionnaire et va rentrer "Cherbourg"

**Comment by pierre-H** [comment]
:smile: mais mami ne va pas gérer le serveur, c'est son petit-fils geek qi va le gérer.
De toutes les façons, on utilise une spinbox qui ne permettra de rentrer que un nombre (ce qui se passe actuellement avec le client).

**Comment by rcharb1** [comment]
alors c'est d'accord pour choisir le port où tournera le serveur!

**Comment by rcharb1** [comment]
Est-ce que tu pourrais me décrire vite fait ton travail ?

**Comment by rcharb1** [comment]
Je pense que je tirerais plus de ton explication synthétique que de me perdre dans les fichiers, ce que je peux aussi faire...

**Comment by pierre-H** [comment]
Pour la version avancée, j'ai du mal à comprendre comment faire pour notre programme : 

> Dans la version plus délicate, votre programme doit recevoir des données de plusieurs sources différentes sans pouvoir prédire d'où les données vont arriver

C'est ce que l'on cherche à faire non ? Le serveur doit pouvoir répondre à n'importe quel client.

**Comment by pierre-H** [comment]
Ok pour l'explication. Je vais uploader un fichier dans doc avec les explications...

**Comment by rcharb1** [comment]
en fait, oui, je pense que c'est plus avancé par rapport à cequelui proposait dans son exemple,mais pour nous, c'est l'utilisation basique.
Mais peut-il gérer deux clients à la fois ?

**Comment by rcharb1** [comment]
En attribuant un port à chacun... oui, comme le chat

**Comment by rcharb1** [comment]
Donc en fait, pas si compliqué que ça !

**Comment by pierre-H** [comment]
Soit un port chacun, soit le même port (plus facile à partager aux utilisateurs) avec une file d'attente vu que le moteur ne peut pas traiter plusieurs requêtes en même temps.

**Comment by rcharb1** [comment]
Oui, mais vu le succès énorme que rencontrera notre programme ;-), je crois qu'il vaut mieux un port pour chacun, non ? C'est plus simple aussi, non ?

**Comment by rcharb1** [comment]
Dans le chat, chaque client est représenté par un QTcpSocket, c'est à dire une autre socket, donc ça correspond plutôt à un port pour chaque client ?
`QTcpSocket *nouveauClient = serveur->nextPendingConnection();`

**Comment by pierre-H** [comment]
Non, c'est même port (50885 dans le chat). C'est ça l'astuce de Qt j'ai l'impression.
Donc une socket par utilisateur, avec une file d'attente pour le moteur.

**Comment by rcharb1** [comment]
Mais apparemment, qt gère tout seul la file d'attente

**Comment by pierre-H** [comment]
La file d'attente pour le serveur oui, mais est-ce qu'on peut l'utiliser pour le moteur ... A vérifier...

**Comment by pierre-H** [comment]
Vu que le moteur met du temps avant de renvoyer une solution...
Pour la queue, la classe QQueue peut nous être utile...

**Comment by rcharb1** [comment]
Que veux-tu dire ? s'il n'y a qu'un seul client qui peut se connecter au serveur à la fois, il n'y a pas ce problème de file d'attente dans le moteur ? Mais si un utilisateur s'endort dans l'utilisation du programme,ça risque de faire attendre tout le monde...

**Comment by pierre-H** [comment]
Moi je pensais au cas d'une entreprise dont 20 employés veulent utiliser le serveur/moteur : il faut bien une file d'attente dans ce cas là. Si l'un d'entre eux s'endort, cela ne change rien : il devrait y avoir 2 parties : 
le serveur qui s'occupe d'ajouter les demandes dans la file d'attente et de renvoyer les réponses du moteur
le moteur qui traite la file d'attente.
Si le moteur travaille, il ne faudrait pas que le serveur bloque les nouvelles demandes.

**Comment by rcharb1** [comment]
ah, j'avais mal compris!

**Comment by pierre-H** [comment]
Pas de souci. C'est à mon avis ça la version avancé du projet...

**Comment by rcharb1** [comment]
et l'utilisateur peut avoir le formulaire devant lui pendant une heure, ce ne sera pas encore considéré comme une demande tant qu'il ne l'a pas envoyée,c'est ça ?

**Comment by pierre-H** [comment]
oui exactement !

**Comment by rcharb1** [comment]
Parfait ! D'autres choses ?

**Comment by pierre-H** [comment]
Le chat fonctionne de la même façon, sauf qu'il n'y a pas de gros moteur.

**Comment by rcharb1** [comment]
J'aurais tellement aimé pouvoir travailler aujourd'hui dessus... Vraiment, ça s'est passé comme un ouragan, on m'a téléphoné "la formation a commencé hier,si tu peux venir passer le test aujourd'hui, tu peux commencer !" et voilà ! maintenant, il faut que je finisse d'étudier le contrat pour être prête à voyager à 7h demain !

**Comment by pierre-H** [comment]
Pour moi, j'ai l'impression qu'on a fait le tour.
Donc, tu essayes de faire ça vendredi, je me trompe ?
Si c'est bien cela, essaye de mettre un maximum de choses dimanche pour que je puisse voir un peu où on en est et peut être faire une réunion dimanche soir. Cela te va ?

**Comment by pierre-H** [comment]
Quel port par défaut ?

**Comment by rcharb1** [comment]
1050 ? Avec un random pour le cas où il est pris ?

**Comment by pierre-H** [comment]
Le client : la saisie de l'IP et du port dans le formulaire de saisie ou dans des fenêtres séparées (solution actuelle) ?

**Comment by rcharb1** [comment]
Oui, je ferai de mon mieux vendredi, en essayant aussi de montrer à la maison que j'existe encore...

**Comment by pierre-H** [comment]
On avais pas mis de côté l'histoire du random pour donner le choix à l'utilisateur ?

**Comment by pierre-H** [comment]
:smile:

**Comment by rcharb1** [comment]
Finalement, c'est très bien comme c'est, avec la fenêtre séparée, c'est plus logique de ne pas pouvoir accéder au formulaire si pas de connexion...

**Comment by rcharb1** [comment]
Donc ici, par port par défaut, tu entends de quel côté : client ou serveur ?

**Comment by pierre-H** [comment]
Ok.

Pour le port : les deux.

**Comment by rcharb1** [comment]
de la marmelade à la place du cerveau, ce n'est pas réputé pour marcher très bien...

**Comment by pierre-H** [comment]
:)

**Comment by rcharb1** [comment]
Donc au choix ou par défaut ? C'est contradictoire, non ?

**Comment by pierre-H** [comment]
Non, par défaut on propose 1050, mais tout en laissant le choix aux utilisateurs de changer cette valeur. Dans la spinBox on aurait donc : default = 1050 mais min = 1024 et max = 65536

**Comment by rcharb1** [comment]
Ah... Je comprends très vite quand on m'explique longtemps !

**Comment by rcharb1** [comment]
Donc allons avec 1050

**Comment by pierre-H** [comment]
:) courage !

**Comment by pierre-H** [comment]
Ok !
On a fait le tour je pense alors

**Comment by rcharb1** [comment]
Merci beaucoup de toute la patience et l'encouragement. De mon côté, je fais de mon mieux pour vendredi et bravo pour le boulot déjà fait !

**Comment by rcharb1** [comment]
parfait, à dimanche soir sûrement !

**Comment by pierre-H** [comment]
Avec plaisir. Bon courage ! Le début n'est pas simple mais après on prend le rythme !
Bon courage !
Je vais mettre dans la soirée l'explication dans le dossier doc en postant un post ici pour que tu reçois un mail dès que c'est fait.
Merci !

A dimanche !

**Comment by rcharb1** [comment]
merci et bonne nuit !

**Comment by pierre-H** [comment]
Bonne nuit !

**Comment by pierre-H** [comment]
Envoyé ! N'hésite pas à me dire si tu as besoin de plus d'explications concernant un certain point.
Il faudrait faire de l'ordre dans le dépot. Fais comme tu le sens (un dossier serveur par exemple). Les rapports de chacun ne sont pas nécessaires...

**Comment by pierre-H** [comment]
Tu peux utiliser la classe **ProtocoleODC_client** pour créer l'équivalente pour le serveur **ProtocoleODC_serveur**

**Comment by rcharb1** [comment]
Je suis en train de travailler dessus.
Pour l'organisation des dossiers, j'ai effectivement fait un dossier réseau dans le dossier général où j'ai mis un dossier Doc, ton dossier client et mon dossier serveur , et pour la classe PreotocoleODC_serveur, je garde exactement ta manière d'écrire.
Bravo pour ce beau boulot !
Je te mets au courant de l'avancée de la journée...

**Comment by rcharb1** [comment]
Voici l'avancée des travaux : 
J'ai d'abord bien lu ce que tu avais ajouté.
Ensuite, je me suis occupée de ranger un peu les dossiers,ça fait plus propre avec la nouvelle organisation.
J'ai écrit la fonction createMessage du serveur, mais je ne l'ai pas encore faite tourner parce que je n'ai pas encore eu le temps de démêler les dizaines de fichiers concernés... Je pensequ'il faut que je fasse un projet Qt spécialement pour le serveur, et que je réussisse à compiler ce que j'ai déjà fait.
Pour l'instant, je m'embrouille pas mal dans tous ces fichiers : lequel est nécessaire, lequel ne l'est pas ? Comment configurer le fichier .pro et autres questions de ce style...
C'est la première fois que je dois me débrouiller dans un projet d'une telle taille (pour le projet original,c'était toi qui t'étais occupé de ça !!)
J'ai aussi écrit une petite doc pour le serveur.
La classe ProtocoleODC_serveur est extrêmement simple puisque le principe de ce moteur est qu'il reçoit en arg un objet MoteurCalculs avec les données et retourne ce même objet modifié avec donc les réponses dedans.
En gros, même si je n'ai pas pu travailler toute la journée comme j'en avais envie, j'ai au moins pu comprendre le code déjà fait, et dessiner les lignes globales de ce que je devais faire...
Encore une fois désolée de te retarder... J'espère pouvoir continuer à avancer le plus rapidement possible,et n'hésite pas à me faire savoir si mon programme impossible te dérange trop et que tu veux être libéré...

**Comment by pierre-H** [comment]
Pour ce qui est de Qt et du fichier pro, Qt Creator le génère tout seul. Il faut juste ajouter les lignes : 
`QT += widgets`
`QT += core`
`QT += network`

Pour le compiler, il n'y a rien à configurer.
Moi aussi j'ai créé un projet séparé pour le client. C'est plus raisonnable.
Penses-tu pouvoir travailler dimanche ?
Bon week-end

**Comment by pierre-H** [comment]
Hello,
Des news ?

**Comment by rcharb1** [comment]
Pas de dimanche en Israel... Je rentre de la formation, je vais faire de mon mieux pour avancer encore un peu ce soir...

**Comment by pierre-H** [comment]
Ah oui j'ai oublié.
Ce serait bien oui. Ce projet est très important car c'est la dernière chose qu'il me reste pour la L2 qui doit être validée cette année pour que je puisse changer de fac.
